suite/tests/client-interface/cleancall-opt-1.dll.c

```diff
@@ -0,0 +1,159 @@
+ * Copyright (c) 2011 Massachusetts Institute of Technology  All rights reserved.
+ * *******************************************************************************/
+
+/*
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * * Redistributions of source code must retain the above copyright notice,
+ *   this list of conditions and the following disclaimer.
+ *
+ * * Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of MIT nor the names of its contributors may be
+ *   used to endorse or promote products derived from this software without
+ *   specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL MIT OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+/* Test the clean call inliner. */
+
+#include "dr_api.h"
+
+#ifdef WINDOWS
+#define BINARY_NAME "client.cleancall-opt-1.exe"
+#else
+#define BINARY_NAME "client.cleancall-opt-1"
+#endif
+
+/* List of instrumentation functions. */
+#define FUNCTIONS() \
+        FUNCTION(modify_gprs) \
+        LAST_FUNCTION()
+
+#include "cleancall-opt-shared.h"
+
+static void event_exit(void);
+static dr_emit_flags_t event_basic_block(void *dc, void *tag, instrlist_t *bb,
+                                         bool for_trace, bool translating);
+
+DR_EXPORT void
+dr_init(client_id_t id)
+{
+    dr_register_exit_event(event_exit);
+    dr_register_bb_event(event_basic_block);
+    dr_fprintf(STDERR, "INIT\n");
+
+    /* Lookup pcs. */
+    lookup_pcs();
+    codegen_instrumentation_funcs();
+}
+
+static void
+event_exit(void)
+{
+    int i;
+    free_instrumentation_funcs();
+
+    for (i = 0; i < N_FUNCS; i++) {
+        DR_ASSERT_MSG(func_called[i],
+                      "Instrumentation function was not called!");
+    }
+    dr_fprintf(STDERR, "PASSED\n");
+}
+
+static dr_emit_flags_t
+event_basic_block(void *dc, void *tag, instrlist_t *bb,
+                  bool for_trace, bool translating)
+{
+    instr_t *entry = instrlist_first(bb);
+    app_pc entry_pc = instr_get_app_pc(entry);
```

Comment:
If you want to get the starting app pc of the basic block, you might want to use app_pc dr_fragment_app_pc, 
http://dynamorio.org/docs/dr__tools_8h.html#a724aa7f7e87a604071fffa0a01925bf0